jfrog-cli/artifactory/commands/curl/curl.go

```diff
@@ -0,0 +1,247 @@
+	"github.com/jfrog/jfrog-client-go/utils/errorutils"
+	"github.com/jfrog/jfrog-client-go/utils/log"
+	"io"
+	"os/exec"
+	"strings"
+)
+
+func Execute(args []string) error {
+	// Get curl execution path.
+	execPath, err := exec.LookPath("curl")
+	if err != nil {
+		return errorutils.CheckError(err)
+	}
+
+	// Create curl command.
+	command := &CurlCommand{ExecutablePath: execPath, Arguments: args}
+
+	// Get server-id, remove from the command if exists.
+	serverIdValue, err := command.getAndRemoveServerIdFromCommand()
+	if err != nil {
+		return err
+	}
+
+	// Get Artifactory details for this ID.
+	artDetails, err := config.GetArtifactorySpecificConfig(serverIdValue)
+	if err != nil {
+		return err
+	}
+
+	// If the command already include credentials flag, return an error.
+	if command.isCredentialsFlagExists() {
+		return errorutils.CheckError(errors.New("Curl command must not include credentials flag (-u or --user)."))
+	}
+
+	// Get target url for the curl command.
+	urlIndex, targetUrl, err := command.buildCommandUrl(artDetails.Url)
+	if err != nil {
+		return err
+	}
+
+	// Replace url argument with complete url.
+	command.Arguments[urlIndex] = targetUrl
+
+	log.Debug(fmt.Sprintf("Executing curl command: '%s -u***:***'", strings.Join(command.Arguments, " ")))
+	// Add credentials flag to Command. In case of flag duplication, the latter is used by Curl.
+	credFlag := fmt.Sprintf("-u%s:%s", artDetails.User, artDetails.Password)
+	command.Arguments = append(command.Arguments, credFlag)
+
+	// Run curl.
+	return gofrogcmd.RunCmd(command)
+}
+
+func (curlCmd *CurlCommand) buildCommandUrl(artifactoryUrl string) (urlIndex int, urlValue string, err error) {
+	// Find command's URL argument.
+	// Representing the target API for the Curl command.
+	urlIndex, urlValue = curlCmd.findUrlValueAndIndex()
+	if urlIndex == -1 {
+		err = errorutils.CheckError(errors.New("Could not find argument in curl command."))
+		return
+	}
+
+	// If user provided full-url, throw an error.
+	if strings.HasPrefix(urlValue, "http://") || strings.HasPrefix(urlValue, "https://") {
+		err = errorutils.CheckError(errors.New("Curl command must not include full-url, provide only the REST API (e.g '/api/system/ping')."))
+		return
+	}
+
+	// Trim '/' prefix if exists.
+	if strings.HasPrefix(urlValue, "/") {
+		urlValue = strings.TrimPrefix(urlValue, "/")
+	}
+
+	// Attach Artifactory's url to the api.
+	urlValue = artifactoryUrl + urlValue
+
+	return
+}
+
+// Get --server-id flag value from the command, and remove it.
+func (curlCmd *CurlCommand) getAndRemoveServerIdFromCommand() (string, error) {
+	// Get server id.
+	serverIdFlagIndex, serverIdValueIndex, serverIdValue, err := curlCmd.findFlag("--server-id")
+	if err != nil {
+		return "", err
+	}
+
+	// Remove --server-id from Command if required.
+	if serverIdFlagIndex != -1 {
+		curlCmd.Arguments = append(curlCmd.Arguments[:serverIdFlagIndex], curlCmd.Arguments[serverIdValueIndex+1:]...)
```

Comment:
Question - couldn't this cause an "out of bound" error, if the --server-id is the last flag?

Reply:
In such case - go allows it and doesn't cause an out of bound error.